### PR TITLE
Feature/open tab update on pull

### PR DESCRIPTION
**Pull request by gunther-bachmann**

### CODE REVIEW

File: src/app/modules/persistence-service/persistence.service.spec.ts [hunk 1]
```diff
@@ -38,7 +40,184 @@ describe('PersistenceService', () => {
...
+      // given
+      persistenceService.startSubscriptions();
+
+      // when
+      messagingService.publish('navigation.open', 'some/file');
+      messagingService.publish('navigation.open', 'some/other/file');
+      messagingService.publish('editor.dirtyStateChanged', { path: 'yet/another', dirty: true });
+      messagingService.publish('editor.dirtyStateChanged', { path: 'and/yet/another', dirty: true });
+      messagingService.publish('editor.dirtyStateChanged', { path: 'and/still/another', dirty: true });
+      messagingService.publish('editor.close', { id: 'some/file' }); // not open not dirty
```
Comment (jjelschen) [inline src/app/modules/persistence-service/persistence.service.spec.ts]: You could use the file paths to express their intended state / use in the test scenario. You'd get rid of the comments, and it would be easier to keep track, and cross-check with the expected result in the _then_ block.

File: src/app/modules/persistence-service/persistence.service.spec.ts [hunk 2]
```diff
@@ -38,7 +40,184 @@ describe('PersistenceService', () => {
...
+      tick();
+
+       // then
+      const pullMatched = httpMock.match(pullMatcher)[0];
+      expect(pullMatched.request.body).toEqual(jasmine.objectContaining(
+        { resources: [ 'some/file', 'some/other/file' ], dirtyResources: [ 'and/yet/another' ] }));
+
+    })));
+
+  it('returns a conflict if backend reports a conflict in the backend', fakeAsync(inject([HttpTestingController, PersistenceService],
```
Comment (jjelschen) [inline src/app/modules/persistence-service/persistence.service.spec.ts]:
```suggestion
  it('returns a conflict if backend reports one', fakeAsync(inject([HttpTestingController, PersistenceService],
```

File: src/app/modules/persistence-service/persistence.service.spec.ts [hunk 3]
```diff
@@ -38,7 +40,184 @@ describe('PersistenceService', () => {
...
+      const message = `The file '${tclFileTarget}' already exists.`;
+      const expectedConflict = new Conflict(message);
+
+      // when
+      persistenceService.copyResource(tclFileTarget, tclFilePath)
+
+      // then
+      .then((result) => {
+        expect(result).toEqual(expectedConflict);
+      }, (response) => {
```
Comment (jjelschen) [inline src/app/modules/persistence-service/persistence.service.spec.ts]:
```suggestion
      }, (error) => {
```
(or "rejected", "exception", etc., just for readability – calling it "response" may be confusing, especially in conjunction with the following test failure message…)

File: src/app/modules/persistence-service/persistence.service.ts [hunk 4]
```diff
@@ -19,59 +24,226 @@ export abstract class AbstractPersistenceService {
...
+  changedResources: Array<string>;
+  backedUpResources: Array<BackupEntry>;
+}
+
 @Injectable()
 export class PersistenceService extends AbstractPersistenceService {
 
   private serviceUrl: string;
   private listFilesUrl: string;
+  private openNonDirtyTabs: Array<string>;
```
Comment (jjelschen) [inline src/app/modules/persistence-service/persistence.service.ts]: Would it be an option to use `Set`s instead? I would imagine it could simplify the code a bit, since you wouldn't have to manually check for duplicates…

File: src/app/modules/persistence-service/persistence.service.ts [hunk 5]
```diff
@@ -19,59 +24,226 @@ export abstract class AbstractPersistenceService {
...
+  private openDirtyTabs: Array<string>;
+  private subscriptions: Subscription[];
+
+  startSubscriptions(): void {
+    this.subscriptions = new Array();
+    this.subscriptions.push(this.messagingService.subscribe(NAVIGATION_OPEN, (id) => {
+      this.removeNonDirtyTab(id);
+      this.openNonDirtyTabs.push(id);
+    }));
+    this.subscriptions.push(this.messagingService.subscribe(NAVIGATION_CLOSE, () => {
```
Comment (jjelschen) [inline src/app/modules/persistence-service/persistence.service.ts]: Just as a side note (I know this event is a preexisting one), `NAVIGATION_CLOSE` suggests its the opposite of `NAVIGATION_OPEN`, but that would actually be `EDITOR_CLOSE`, whereas `NAVIGATION_CLOSE` is published on logout… :confused: 
Not part of this PR, but we might want to consider renaming that event at some point.

File: src/app/modules/persistence-service/persistence.service.ts [hunk 6]
```diff
@@ -19,59 +24,226 @@ export abstract class AbstractPersistenceService {
...
+    const dirtyTab = this.openDirtyTabs.indexOf(path);
+    if (dirtyTab >= 0) {
+      this.openDirtyTabs.splice(dirtyTab, 1);
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+  private removeNonDirtyTab(path: string): boolean {
```
Comment (jjelschen) [inline src/app/modules/persistence-service/persistence.service.ts]: code clone…

File: src/app/modules/persistence-service/persistence.service.ts (hunk 5 above)
Comment (gunther-bachmann) [reply src/app/modules/persistence-service/persistence.service.ts]: right, i'll open a fix for that

File: src/app/modules/persistence-service/persistence.service.ts (hunk 4 above)
Comment (gunther-bachmann) [reply src/app/modules/persistence-service/persistence.service.ts]: much better

File: src/app/modules/persistence-service/persistence.service.ts (hunk 6 above)
Comment (gunther-bachmann) [reply src/app/modules/persistence-service/persistence.service.ts]: no longer necessary
